README


This archive contains the distribution Test-CleanNamespaces,
version 0.16:

  Check for uncleaned imports

This software is copyright (c) 2009 by Florian Ragwitz.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.020.



Changes
Revision history for Test-CleanNamespaces

0.16      2014-08-28 23:55:47Z
  - skip Moose-related tests for normal installs, to get out of circularity
    hell if Moose is installed but broken and needing an upgrade

0.15      2014-08-16 00:18:15Z
  - bump Package::Stash prereq to ensure used methods are available

0.14      2014-08-11 21:44:34Z
  - remove accidental dependency on Class::MOP (which turned into a circular
    dependency with Moose-2.1211)

0.13      2014-06-24 19:14:03Z
  - bump (optional) prereq on Moo to get some fixes for handling roles

0.12      2014-06-18 01:58:18Z
  - also special case constant subs, which show up in the symbol table before
  perl 5.010 as 'constant::__ANON__'

0.11      2014-06-15 04:02:16Z
  - the return value of namespaces_clean() and all_namespaces_clean() is now
    consistent with the result of the test(s).
  - fixed tests to expose an issue with Mouse classes, and documented this
    (now) known issue

0.10      2014-05-12 04:30:17Z (TRIAL RELEASE)
  - fix test diagnostics in the success case
  - fix use of a new Scalar::Util export in tests

0.09      2014-05-07 05:06:58Z (TRIAL RELEASE)
  - increased version for optional Role::Tiny prereq, to get the needed
    is_role interface
  - now ignoring 'import', 'unimport' subs; also handle users of
    Moose::Exporter without erroring
  - now ignoring overloaded methods
  - now also properly handle classes using Class::MOP directly

0.08      2014-03-09 01:08:34Z
  - fixed method identification in Moose and Mouse roles

0.07      2014-03-02 06:55:41Z
  - remove test prereqs that are meant to be optional (added in v0.06)
  - fix handling of subs composed from a role (Moose::Role, Moo::Role,
    Mouse::Role, and Role::Tiny), broken in v0.06

0.06      2014-03-01 19:14:39Z
  - lots more tests
  - in failing tests, identify the source of the uncleaned sub(s)
  - remove the dependency on Class::MOP by inspecting the package stash
    directly
 - line numbers in shipped code are now almost the same (within 3) as the
   repository source, for easier debugging

0.05      2013-09-25 03:34:13Z
  - re-release with fixed compile test

0.04      2013-09-14 17:32:26Z
  - remove use of deprecated Class::MOP::load_class

0.03   2010-05-07 22:29:00 +0200
  - Stop using the deprecated get_method_map method.

0.02   2009-05-28 06:05:26 +0200
  - Declare dependency on Test::Tester.

0.01   2009-05-16 18:32:11 +0200
  - Initial release.


t/lib/Role.pm
use strict;
use warnings;
package Role;

use Role::Tiny 1.003000;    # for is_role, used by our heuristic
use Scalar::Util 'reftype';
use namespace::clean;

sub role_stuff {}

1;


t/lib/Clean.pm
use strict;
use warnings;
package Clean;

use ExporterModule qw/stuff/;
use Scalar::Util 'refaddr';
use namespace::clean;

sub method { }

sub callstuff { stuff(); 'called stuff' }

1;


t/lib/Dirty.pm
use strict;
use warnings;
package Dirty;

use ExporterModule qw/stuff/;

sub method { }

sub callstuff { stuff(); 'called stuff' }

1;


inc/DynamicPrereqs.pm
use strict;
use warnings;
package inc::DynamicPrereqs;

use Moose;
with 'Dist::Zilla::Role::InstallTool', 'Dist::Zilla::Role::MetaProvider';
use List::Util 'first';

# this is a horrible hack and I'm doing it much better in its own dist. this
# is very temporary...

sub metadata { return +{ dynamic_config => 1 } }

sub setup_installer
{
    my $self = shift;

    my $file = first { $_->name eq 'Makefile.PL' } @{$self->zilla->files};
    my $content = $file->content;

    $self->log_debug('Inserting dynamic prereq into Makefile.PL...');
    $self->log_fatal('failed to find position in Makefile.PL to munge!')
        if $content !~ m'^my %FallbackPrereqs = \(\n(?:[^;]+)^\);$'mg;

    my $pos = pos($content);

    $content = substr($content, 0, $pos)
        . "\n\n"
        . q|$WriteMakefileArgs{PREREQ_PM}{'Role::Tiny'} = $FallbackPrereqs{'Role::Tiny'} = '1.003000' if eval { require Role::Tiny; 1 };|
        . "\n" . substr($content, $pos, -1);

    $file->content($content);
    return;
}

1;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Check for uncleaned imports",
  "AUTHOR" => "Florian Ragwitz <rafl\@debian.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Test-CleanNamespaces",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Test::CleanNamespaces",
  "PREREQ_PM" => {
    "File::Find::Rule" => 0,
    "File::Find::Rule::Perl" => 0,
    "File::Spec::Functions" => 0,
    "Module::Runtime" => 0,
    "Package::Stash" => "0.14",
    "Sub::Exporter" => 0,
    "Sub::Identify" => 0,
    "Test::Builder" => 0,
    "namespace::clean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Exporter" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Module::Runtime" => 0,
    "Scalar::Util" => 0,
    "Test::Deep" => 0,
    "Test::More" => "0.88",
    "Test::Requires" => 0,
    "Test::Tester" => 0,
    "Test::Warnings" => "0.009",
    "constant" => 0,
    "if" => 0,
    "lib" => 0,
    "overload" => 0,
    "parent" => 0
  },
  "VERSION" => "0.16",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Find::Rule" => 0,
  "File::Find::Rule::Perl" => 0,
  "File::Spec" => 0,
  "File::Spec::Functions" => 0,
  "Module::Runtime" => 0,
  "Package::Stash" => "0.14",
  "Scalar::Util" => 0,
  "Sub::Exporter" => 0,
  "Sub::Identify" => 0,
  "Test::Builder" => 0,
  "Test::Deep" => 0,
  "Test::More" => "0.88",
  "Test::Requires" => 0,
  "Test::Tester" => 0,
  "Test::Warnings" => "0.009",
  "constant" => 0,
  "if" => 0,
  "lib" => 0,
  "namespace::clean" => 0,
  "overload" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);

$WriteMakefileArgs{PREREQ_PM}{'Role::Tiny'} = $FallbackPrereqs{'Role::Tiny'} = '1.003000'
if eval { require Role::Tiny; 1 };



unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


